### ISSUE TITLE
Htsjdk defines queryname sorting differently than samtools which can cause failures

### ISSUE THREAD

**Issue by lbergelson** [samtools/htsjdk#766]
samtools queryname sort (`samtools sort -n`) sorts readnames in a numerically aware fashion
htsjdk expects queryname sorting to be done by lexicographical comparison of readnames

This causes  reads like HXXX:999 and HXXX:1000 to sort in different ways when using picard SortSam vs samtools to sort.

This creates problems when reading and then writing files sorted with samtools.  If a vcf writer is set to assume sorted, than it will fail validation checks when outputting a file that was sorted in samtools order.

BAM spec doesn't define which order is correct, so it would be good if we could accept either.  We could potentially broaden how we interpret queryname order in order to allow either sorting.

**Comment by droazen** [comment]
A big :+1: to being able to accept either order -- I think this has been a pain point for many years. @vdauwera care to comment?

**Comment by nh13** [comment]
@lbergelson @droazen make sure to do a thorough search of the samtools mailing list since this has been raised many, many times ([example](https://sourceforge.net/p/samtools/mailman/samtools-devel/thread/509CF4E2-2AB8-450C-9B6A-5487B2E76B15%40sanger.ac.uk/#msg26674128)).  I would even suggest bringing back up the topic to the samtools-devel list, as folks may not see this issue.

Also, why would the read names from a SAM or BAM make an impact on a VCF writer?

**Comment by vdauwera** [comment]
+1 for accepting either

**Comment by droazen** [comment]
@nh13 We're aware that this has been discussed in the past -- but nothing has come of those discussions. We think this is a bug in htsjdk. If both definitions of queryname ordering are compliant with the BAM spec, then it seems to us that htsjdk should accept both when doing validation.

Note that we're not proposing standardizing on a single universal order (which is what the thread you linked to contemplates), we just want to relax the validation that is done in htsjdk to be able to accept both conventions.

**Comment by nh13** [comment]
Do folks rely on a specific sort order implementation?  If so, then validating that the input follows htsjdk's qname sort order implementation sounds useful to have in the API.

**Comment by droazen** [comment]
Perhaps, but by default htsjdk should not be blowing up on input that is compliant with the spec. Ie., when you tell htsjdk "please check whether the input is not actually queryname-sorted", it should not explode on input that is in fact queryname-sorted according to the spec as it does now.

**Comment by lbergelson** [comment]
@nh13 vcf writer -> bam writer.  Mental failure due to sleep deprivation.

**Comment by nh13** [comment]
I am not saying "No" (or saying "Yes") just that it isn't as simple as "the spec says...".  I believe that the SamFileValidator already is stricter than the spec: https://broadinstitute.github.io/picard/sam-differences.html. 

If you want to relax the validation of queryname sort order, how would that be done?  It seems like we would have to just skip sort order validation if the input is queryname, or perhaps are you saying you would like have all queryname sort orders in the wild implemented in htsjdk and we figure out on the fly which one it is and validate against that?

**Comment by lbergelson** [comment]
Does anyone depend on the specifics of the queryname sort?  We don't actually know.  What order do fastq's come from the sequencers.  Does it match either sort?  

As far as I know, the only thing important about queryname sort is that pairs are together.  It's just hard to validate that pairs are together if you can't agree on a unique sorting.

**Comment by lbergelson** [comment]
It's definitely the case that SamFileValidator makes additional assumptions on top of the spec.  It's just that if possible, it would be good to accept files sorted in any valid queryname sort if theres a way to do so safely.

**Comment by droazen** [comment]
@nh13 I'm saying we should do what is done for validation of coordinate-sorted input -- use a more relaxed ordering than the htsjdk comparators impose when validating that a particular ordering exists.

Eg., `SAMRecordCoordinateComparator` imposes all sorts of crazy tie-breaking rules when there are two records with the same position, but we don't require that these tie-breaking rules have been applied when checking ordering in, eg., `SamFileWriterImpl.assertPresorted()`. It should be possible to do something similar for queryname-sorted data.

**Comment by yfarjoun** [comment]
The non-lexicographic sort-order seems to me very complex and not even well
defined.  I prefer to say "this is not queryname sorted" and ask that the
input be resorted before processing with htsjdk than to build a complex
sort function based on an arbitrary implementation we do not control and
lose the concept of queryname sorted (PIcard style) being a unique thing.

:-1: from me.

**Comment by droazen** [comment]
@yfarjoun It doesn't need to be complex, or a full re-implementation of the logic from samtools -- we just need to verify that identical read names are grouped together, and ignore all other arbitrary conventions imposed by the various htsjdk/samtools comparators. Right now, we blow up by default on an entire class of valid bams produced by a very popular set of tools -- don't you think we should try to fix this?

**Comment by yfarjoun** [comment]
I guess I'm not sure that I would consider them valid.

While the sam spec doesn't specify it, My interpretation is that a string
(which is what QNAME is) has one "natural" sort order and that is
lexicographic. any other order needs to be specifically defined and put
into the spec.

You are talking about being queryname *GO (group order)* which is a concept
that exists in the sam spec and we could implement validating that using a
hash-set to check that every queryname show up in a cluster an no-where
else. the only problem is that these "invalid" bams are not headed as
having a group-order. We could write a GO validator and have an option to
ignore the header's SO and use a commandline provided order. in this case
we will use GO=queryname in the commandline...

I'd support that kind of solution.

**Comment by tfenne** [comment]
@droazen The problem with `we just need to verify that identical read names are grouped together` is that you need either time or memory in proportion to the number of reads.  In many cases you simply don't know how many reads there are per template, so the only way to validate a group ordering _without_ validating a sort ordering is to keep track of all the querynames seen so far and ensure that the same query name isn't seen twice with one or more other querynames intervening.

I think the only way to do what you want is to implement a comparator for the samtools-style queryname sort, which is currently underspecified.  I see the value in what you're proposing, but I think if you want to be able to validate/accept that style of sorting in samtools then the sort algorithm needs to be specified clearly somewhere.  For example, how should various symbols be treated if found between alpha and numeric segments like `abc-123` or `abc.123`?  How should you interpret numbers with decimals?  With multiple decimal points in them `123.456.789`.  Etc.

**Comment by d-cameron** [comment]
> the only way to validate a group ordering without validating a sort ordering is to keep track of all the querynames

If the above is unpalatable then I'd setting for an option for approximate validation such as a sliding window of names of the last of the last X million reads. If memory is problematic, just store a unique hash of the read names. A 128 bit or larger hash would avoid both the birthday problem and having to store the actual read names.

The two key scenarios one encounters when wanting name grouped bams are coordinate sorted bams, and raw aligner output bams. The sliding window approach would cover both these scenarios as paired-end reads are nearby but not adjacent in coordinate sorted bams (for both read pair and mate pair libraries), and raw aligner output misorderings are from output file race conditions and are also nearby. This approach might not pick up multi-mapped single-ended reads but it'd be a whole lot better than the current state of affairs in which SAM ambiguities remain indefinitely because resolving it is a breaking change for someone.

**Comment by tfenne** [comment]
Agreed @d-cameron, there are approximate or limited solutions that require less memory.  My point was simply that validating query grouped isn't a trivial undertaking, and that most solutions will add significant overhead in either compute or memory.

**Comment by yfarjoun** [comment]
I agree that one could implement a solution that validates group order. 
I disagree that the queryname order should be more permissive.

Perhaps one could use a hashed approach and store as value the position of the read in the file. Then in the case of a collision (one will happen eventually), a few random-access reads of the bam will discern if the collision was real or due to a hash-collision. This of-course needs us to be able to access an unsorted bam using random access which I'm not sure we can....
